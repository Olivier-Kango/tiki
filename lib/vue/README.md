# lib/vue/

Historical vuejs integration.  The goal is to move everything to vue 3.

As of 2023-04-11, what's left is tiki field rules, and ui-predicate-vue