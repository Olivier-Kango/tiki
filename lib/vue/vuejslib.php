<?php
// (c) Copyright by authors of the Tiki Wiki CMS Groupware Project
//
// All Rights Reserved. See copyright.txt for details and a complete list of authors.
// Licensed under the GNU LESSER GENERAL PUBLIC LICENSE. See license.txt for details.
// $Id$


class VueJsLib
{
	/**
	 * @param string $str    body of the vue document
	 * @param string $name   name of the component
	 * @param bool   $app    whether to create the App
	 * @param array  $data   values to expose to the Vue App
	 * @param bool   $minify or not
	 *
	 * @return string
	 * @throws Exception
	 */

	public function processVue($str, $name = '', $app = false, $data = [], $minify = false)
	{
		$headerlib = TikiLib::lib('header');

		if (is_readable($str)) {
			$str = file_get_contents($str);
		}

		// process some shorthand syntax that doesn't work for us
		$str = preg_replace('/\s(@)(?=' . implode('|', $this->jsEvents()) . ')\b/', ' v-on:', $str);

		$dom = new DOMDocument('1.0', 'UTF-8');
		$dom->loadHTML("<html lang=\"en\"><body>$str</body></html>");

		$script = $dom->getElementsByTagName('script');
		$template = $dom->getElementsByTagName('template');
		$style = $dom->getElementsByTagName('style');

		if (! $name && $app) {
			$name = 'App';
		}

		$nameLowerCase = strtolower($name);

		if (! $name && $app) {
			$name = 'App';
		}

		$nameLowerCase = strtolower($name);

		if ($script->length) {    // required
			$javascript = $script[0]->nodeValue;
			preg_match('/export default {(.*)}/ms', $javascript, $match);

			if ($match) {
				$originalExport = $export = $match[1];
				if ($template->length) {
					$templateNode = $template[0];
					$export .= ', template: `' . $this->getInnerHtml($templateNode) . '`';
					$javascript = str_replace($originalExport, $export, $javascript);
				}
				//$headerlib->add_js_module($javascript);
				// embedded modules cannot export apparently, also can't be found by import fns

			}
			global $tikidomainslash;
			$tempDir = './temp/public/' . $tikidomainslash;
			$hash = $nameLowerCase ? $nameLowerCase : md5(serialize($javascript));

			$file = $tempDir . "vue_" . $hash . ".js";
			if ($minify) {
				$minifier = new MatthiasMullie\Minify\JS($javascript);
				$minifier->minify($file);
			} else {
				file_put_contents($file, $javascript);
			}
			chmod($file, 0644);

			if ($app) {
				$data = json_encode($data);

				$headerlib->add_js_module(
					"
import $name from \"$file\";

var vm = new Vue({
	  render: h => h($name),
	  data: function () { return $data; },
	}).\$mount(`#$nameLowerCase`);
"
				);
				return "<div id=\"$nameLowerCase\"></div>";
			}
		}

		return '';
	}

	// thanks dpetroff https://www.php.net/manual/en/class.domelement.php#101243
	function getInnerHtml($node)
	{
		$innerHTML = '';
		$children = $node->childNodes;
		foreach ($children as $child) {
			$innerHTML .= $child->ownerDocument->saveXML($child);
		}

		return str_replace('&#13;', "\r", $innerHTML);
	}

	/**
	 * generates a predicate ui vue.js based rules gui for a tracker field
	 *
	 * @param array $params
	 *
	 * @return string
	 * @throws Exception
	 */
	public function getFieldRules($params)
	{
		if (empty($params['fieldId'])) {
			Feedback::error(tr('No fieldId for Field Rules'));
		}

		foreach ($params['targetFields'] as & $field) {
			switch ($field['type']) {
				case 'f':    // datetime
				case 'j':    // datepicker
				case 'CAL':  // calendar item
					$field['argumentType'] = 'DateTime';
					break;
				case 'n':    // number
				case 'b':    // currency
					$field['argumentType'] = 'Number';
					break;
				case 'c':    // checkbox
					$field['argumentType'] = 'Boolean';
					break;
				case 'e':    // Category
				case 'd':    // DropDown
				case 'D':    // DropDown with Other
				case 'M':    // Multiselect
				case 'w':    // DynamicList
					$field['argumentType'] = 'Array';
					break;
				default:
					$field['argumentType'] = 'Text';
					break;
			}
		}

		if (is_string($params['rules'])) {
			$params['rules'] = json_decode(html_entity_decode($params['rules']));
		}

		if (! is_object($params['rules']) || empty($params['rules'])) {
			$params['rules'] = [
				'conditions' => null,
				'actions' => null,
				'else' => null,
			];
		}

		$params['definitiion'] = \Tracker\Rule\Definition::get();

		$appHtml = $this->processVue('lib/vue/rules/TrackerRulesApp.vue', 'TrackerRulesApp', true, $params);

		$appHtml .= $this->processVue('lib/vue/rules/TextArgument.vue', 'TextArgument');
		$appHtml .= $this->processVue('lib/vue/rules/NumberArgument.vue', 'NumberArgument');
		$appHtml .= $this->processVue('lib/vue/rules/DateArgument.vue', 'DateArgument');
		$appHtml .= $this->processVue('lib/vue/rules/NoArgument.vue', 'NoArgument');
		$appHtml .= $this->processVue('lib/vue/rules/BoolArgument.vue', 'BoolArgument');

		$appHtml .= $this->processVue('lib/vue/rules/TrackerRules.vue', 'TrackerRules');

		return $appHtml;
	}

	public function generateTrackerRulesJS($fields, $insPrefix = 'ins_', $parentSelector = '.form-group:first') {

		$js = '';

		foreach ($fields as $field) {
			if (! empty( $field['rules'])) {
				$rules = Tiki\Lib\core\Tracker\Rule\Rules::fromData($field['fieldId'], $field['rules']);
				$js .= $rules->getJavaScript($field['fieldId'], $parentSelector);
			}
		}

		return $js;
	}

	/**
	 * @return array
	 */
	private function jsEvents(): array
	{
		return [
			'abort',
			'afterprint',
			'animationcancel',
			'animationend',
			'animationiteration',
			'animationstart',
			'audioprocess',
			'auxclick',
			'beforeprint',
			'beforeunload',
			'blur',
			'canplay',
			'canplaythrough',
			'change',
			'click',
			'close',
			'complete',
			'compositionend',
			'compositionstart',
			'compositionupdate',
			'contextmenu',
			'copy',
			'cut',
			'dblclick',
			'drag',
			'dragend',
			'dragenter',
			'dragleave',
			'dragover',
			'dragstart',
			'drop',
			'durationchange',
			'emptied',
			'ended',
			'error',
			'focus',
			'fullscreenchange',
			'fullscreenerror',
			'keydown',
			'keypress',
			'keyup',
			'load',
			'load',
			'loadeddata',
			'loadedmetadata',
			'loadend',
			'loadstart',
			'message',
			'mousedown',
			'mouseenter',
			'mouseleave',
			'mousemove',
			'mouseout',
			'mouseover',
			'mouseup',
			'offline',
			'online',
			'open',
			'pagehide',
			'pageshow',
			'paste',
			'pause',
			'play',
			'playing',
			'pointerlockchange',
			'pointerlockerror',
			'popstate',
			'progress',
			'ratechange',
			'reset',
			'resize',
			'scroll',
			'seeked',
			'seeking',
			'select',
			'stalled',
			'submit',
			'suspend',
			'timeout',
			'timeupdate',
			'transitioncancel',
			'transitionend',
			'transitionrun',
			'transitionstart',
			'unload',
			'volumechange',
			'waiting',
			'wheel',
			// seems to be a custom vue or ui-predicate event?
			'initialize',
		];
	}

}