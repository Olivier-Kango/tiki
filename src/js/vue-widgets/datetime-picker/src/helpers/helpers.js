import { format } from "date-fns";
import moment from "moment-timezone";

/**
 * @param {Number} date timestamp
 * @param {boolean} useBrowserTimezone
 * @param {string} customTimezone
 * @returns {string}
 * @description
 * Returns the Unix timestamp of the given date.
 */
export function convertToUnixTimestamp(date, useBrowserTimezone, customTimezone) {
    if (useBrowserTimezone) {
        return moment(date).unix();
    }

    const browserOffset = new Date().getTimezoneOffset() * 60;
    const selectedTzOffset = moment.tz(customTimezone).utcOffset() * 60;
    return moment(date).unix() + browserOffset + selectedTzOffset;
}

/**
 * @param {Date|Date[]} date
 * @param {boolean} withTime
 * @param {Locale} locale
 * @returns {string}
 * @description
 * Returns the formatted date string.
 */
export function formatDate(value, withTime, locale) {
    const formatValue = (date) => format(date, withTime ? "Pp" : "P", { locale });
    return Array.isArray(value) ? value.map(formatValue).join(" - ") : formatValue(value);
}

/**
 * @param {Date|Date[]|null} updatedData
 * @param {string} goToURLOnChange
 * @param {number} unixTimestamp
 * @param {number} toUnixTimestamp
 * @param {string} selectedTz
 * @param {string|undefined} globalCallback
 * @returns {void}
 * @description
 * Called when there is the need to navigate to a given URL when the datetime is changed instead of updating the model value.
 */
export function goToURLWithData(updatedData, goToURLOnChange, unixTimestamp, toUnixTimestamp, selectedTz, globalCallback) {
    if (goToURLOnChange) {
        const url = new URL(goToURLOnChange, window.location.origin);
        if (updatedData) {
            url.searchParams.set("todate", unixTimestamp);
            url.searchParams.set("enddate", toUnixTimestamp);
            url.searchParams.set("tzname", selectedTz);
            url.searchParams.set("tzoffset", moment.tz(selectedTz).utcOffset());
        } else {
            url.searchParams.delete("todate");
            url.searchParams.delete("enddate");
            url.searchParams.delete("tzname");
            url.searchParams.delete("tzoffset");
        }
        window.location.href = url;
    } else if (globalCallback && window[globalCallback]) {
        window[globalCallback]({
            date: unixTimestamp,
            enddate: toUnixTimestamp,
            tzname: selectedTz,
            tzoffset: moment.tz(selectedTz.offset).utcOffset(),
        });
    }
}
