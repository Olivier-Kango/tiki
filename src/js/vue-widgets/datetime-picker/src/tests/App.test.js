import { render, screen, fireEvent, within } from "@testing-library/vue";
import VueDatePicker from "@vuepic/vue-datepicker";
import moment from "moment";
import { describe, test, expect, vi, beforeEach } from "vitest";
import { h } from "vue";
import DatetimePicker, { DATA_TEST_ID, TEXT } from "../App.vue";
import * as Helpers from "../helpers/helpers";

vi.mock("@vuepic/vue-datepicker", async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        default: vi.fn().mockImplementation(() => h("div", { "data-testid": DATA_TEST_ID.DATE_PICKER })),
    };
});

vi.mock("../helpers/helpers", async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        convertToUnixTimestamp: vi.fn(),
        formatDate: vi.fn(),
        goToURLWithData: vi.fn(),
    };
});

describe("DatetimePicker", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    describe("render tests", () => {
        test("renders correctly with the given props", () => {
            const givenProps = {
                themeCss: "themes/default/css/default.css",
                inputName: "date",
                timestamp: "10",
                enableTimezonePicker: 1,
                enableTimePicker: 1,
                selectText: "Select",
                cancelText: "Cancel",
                timezone: "America/New_York",
                language: "en",
                rangePicker: true,
                toTimestamp: "20",
                timezoneInput: document.createElement("input"),
                dateTimeInput: document.createElement("input"),
                toDateTimeInput: document.createElement("input"),
            };

            // GIVEN that the call to convertToUnixTimestamp function will return some value
            const expectedConvertedTimestamp = "foo";
            vi.spyOn(Helpers, "convertToUnixTimestamp").mockReturnValue(expectedConvertedTimestamp);

            render(DatetimePicker, {
                props: {
                    ...givenProps,
                },
            });

            const expectedCssThemeLinkElement = screen.getByTestId(DATA_TEST_ID.CSS_THEME);
            expect(expectedCssThemeLinkElement).to.exist;
            expect(expectedCssThemeLinkElement.getAttribute("href")).toBe(givenProps.themeCss);

            const containerElement = screen.getByTestId(DATA_TEST_ID.CONTAINER);
            expect(containerElement).to.exist;

            const datePickerElement = within(containerElement).getByTestId(DATA_TEST_ID.DATE_PICKER);
            expect(datePickerElement).to.exist;
            expect(VueDatePicker).toHaveBeenCalledWith(
                expect.objectContaining({
                    modelValue: [givenProps.timestamp * 1000, givenProps.toTimestamp * 1000],
                    cancelText: givenProps.cancelText,
                    selectText: givenProps.selectText,
                    timezone: { timezone: givenProps.timezone },
                    range: givenProps.rangePicker,
                    locale: givenProps.language,
                    "enable-time-picker": givenProps.enableTimezonePicker,
                }),
                null
            );

            expect(Helpers.convertToUnixTimestamp).toHaveBeenCalledWith(
                givenProps.timestamp * 1000,
                Boolean(givenProps.enableTimezonePicker && !givenProps.timezone),
                givenProps.timezone
            );
            // expect the hidden input value to be whatever the convertToUnixTimestamp function returned
            expect(givenProps.dateTimeInput.value).toBe(expectedConvertedTimestamp);

            expect(Helpers.convertToUnixTimestamp).toHaveBeenCalledWith(
                givenProps.toTimestamp * 1000,
                Boolean(givenProps.enableTimezonePicker && !givenProps.timezone),
                givenProps.timezone
            );
            expect(givenProps.toDateTimeInput.value).toBe(expectedConvertedTimestamp);
            // convertToUnixTimestamp should have been called twice by now
            expect(Helpers.convertToUnixTimestamp).toHaveBeenCalledTimes(2);

            const timezoneContainerElement = within(containerElement).queryByTestId(DATA_TEST_ID.TIMEZONE_CONTAINER);
            expect(timezoneContainerElement).to.exist;

            const timezoneContainerLabelElement = within(timezoneContainerElement).getByTestId(DATA_TEST_ID.LABEL_TIMEZONE_CONTAINER);
            expect(timezoneContainerLabelElement).to.exist;
            expect(timezoneContainerLabelElement.textContent).toBe(TEXT.LABEL_TIMEZONE_CONTAINER);

            const timezoneSelectElement = within(timezoneContainerElement).getByTestId(DATA_TEST_ID.TIMEZONE_SELECT);
            expect(timezoneSelectElement).to.exist;
            const timezones = moment.tz.names();
            expect(timezoneSelectElement.children).toHaveLength(timezones.length);
            timezones.forEach((timezone, index) => {
                const optionElement = timezoneSelectElement.children[index];
                expect(optionElement.value).toBe(timezone);
                expect(optionElement.textContent).toEqual(expect.stringContaining(timezone));
            });

            // snapshot testing
            expect(containerElement).toMatchSnapshot();
        });

        test("renders correctly when optional props are not provided", () => {
            const givenProps = {
                dateTimeInput: document.createElement("input"),
            };

            render(DatetimePicker, {
                props: {
                    ...givenProps,
                },
            });

            expect(VueDatePicker).toHaveBeenCalledWith(
                expect.objectContaining({
                    cancelText: TEXT.LABEL_DEFAULT_CANCEL_BUTTON,
                    selectText: TEXT.LABEL_DEFAULT_SELECT_BUTTON,
                    range: false,
                    locale: "en",
                }),
                null
            );

            const containerElement = screen.getByTestId(DATA_TEST_ID.CONTAINER);
            expect(containerElement).to.exist;

            const timezoneContainerElement = within(containerElement).queryByTestId(DATA_TEST_ID.TIMEZONE_CONTAINER);
            expect(timezoneContainerElement).toBeNull();

            // snapshot testing
            expect(containerElement).toMatchSnapshot();
        });
    });

    describe("Action tests", () => {
        test.each([
            ["the range picker is enabled", [new Date("2021-01-01"), new Date("2021-01-02")], true],
            ["the range picker is disabled", [new Date("2021-01-01")], false],
        ])(
            "should update the hidden timestamp inputs when the date picker value changes and %s and call the emitValueChange prop",
            async (_, givenUpdatedDates, rangePicker) => {
                VueDatePicker = {
                    emits: ["update:modelValue"],
                    setup(props, { emit }) {
                        const handleClick = () => emit("update:modelValue", givenUpdatedDates);
                        return () => h("div", {}, h("button", { onClick: handleClick }, "Select date"));
                    },
                };
                const givenProps = {
                    emitValueChange: vi.fn(),
                    dateTimeInput: document.createElement("input"),
                    toDateTimeInput: document.createElement("input"),
                    rangePicker,
                };
                render(DatetimePicker, {
                    props: givenProps,
                });

                const pickDateButton = screen.getByText("Select date");
                await fireEvent.click(pickDateButton);

                givenUpdatedDates.forEach((date, index) => {
                    expect(Helpers.convertToUnixTimestamp).toHaveBeenCalledWith(date, expect.anything(), expect.anything());
                    expect(givenProps.emitValueChange).toHaveBeenCalledWith({
                        value: expect.anything(),
                        unixTimestamp: Helpers.convertToUnixTimestamp.mock.results[index].value,
                        toUnixTimestamp: Helpers.convertToUnixTimestamp.mock.results[index].value,
                        timezone: expect.anything(),
                    });
                });

                expect(givenProps.dateTimeInput.value).toBe(Helpers.convertToUnixTimestamp.mock.results[0].value);
                if (rangePicker) {
                    expect(givenProps.toDateTimeInput.value).toBe(Helpers.convertToUnixTimestamp.mock.results[1].value);
                }
            }
        );

        test.each([["goToUrlOnChange"], ["globalCallback"]])(
            "should call the goToURLWithData helper function when the date picker value changes and the prop %s is set",
            async (prop) => {
                const givenUpdatedDate = new Date("2021-01-01");
                const givenProps = {
                    [prop]: "foo",
                    timestamp: "10",
                    toTimestamp: "20",
                    timezone: "America/New_York",
                    dateTimeInput: document.createElement("input"),
                };
                VueDatePicker = {
                    emits: ["update:modelValue"],
                    setup(props, { emit }) {
                        const handleClick = () => emit("update:modelValue", [givenUpdatedDate]);
                        return () => h("div", {}, h("button", { onClick: handleClick }, "Select date"));
                    },
                };
                render(DatetimePicker, {
                    props: givenProps,
                });

                const pickDateButton = screen.getByText("Select date");
                await fireEvent.click(pickDateButton);

                expect(Helpers.goToURLWithData).toHaveBeenCalledWith(
                    [givenUpdatedDate],
                    givenProps.goToUrlOnChange,
                    expect.anything(),
                    expect.anything(),
                    expect.anything(),
                    givenProps.globalCallback
                );
            }
        );

        test("should render the updated date when the timezone select value changes and call the emitValueChange prop", async () => {
            VueDatePicker = vi.fn();
            const givenProps = {
                enableTimezonePicker: 1,
                timezone: "foo",
                timezoneFieldName: "timezone",
                emitValueChange: vi.fn(),
                dateTimeInput: document.createElement("input"),
                timezoneInput: document.createElement("input"),
            };

            render(DatetimePicker, {
                props: givenProps,
            });

            expect(VueDatePicker).toHaveBeenCalledWith(expect.objectContaining({ timezone: { timezone: givenProps.timezone } }), null);

            const timezoneSelectElement = screen.getByTestId(DATA_TEST_ID.TIMEZONE_SELECT);
            const selectedTimezone = timezoneSelectElement.children[1].value;
            await fireEvent.update(timezoneSelectElement, selectedTimezone);

            expect(VueDatePicker).toHaveBeenCalledWith(expect.objectContaining({ timezone: { timezone: selectedTimezone } }), null);
            expect(Helpers.convertToUnixTimestamp).toHaveBeenCalledWith(expect.anything(), expect.anything(), selectedTimezone);
            expect(givenProps.emitValueChange).toHaveBeenCalledWith({
                value: expect.anything(),
                unixTimestamp: expect.anything(),
                toUnixTimestamp: expect.anything(),
                timezone: selectedTimezone,
            });
        });
    });
});
