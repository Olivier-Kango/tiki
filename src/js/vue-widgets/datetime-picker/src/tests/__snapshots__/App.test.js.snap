// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatetimePicker > render tests > renders correctly the today shortcut when the range picker is disabled 1`] = `
<div
  class="m-2"
  data-testid="shortcut-today-1720000034990"
>
  <button
    class="btn btn-outline-dark btn-sm w-100"
    data-testid="shortcut-today-from-1720000034990"
    type="button"
  >
    Today
  </button>
</div>
`;

exports[`DatetimePicker > render tests > renders correctly the today shortcut when the range picker is enabled 1`] = `
<div
  class="m-2"
  data-testid="shortcut-today-1720000034990"
>
  <div
    class="d-flex justify-content-between"
  >
    <button
      class="btn btn-outline-dark btn-sm w-100 me-1"
      data-testid="shortcut-today-from-1720000034990"
      type="button"
    >
      Today
    </button>
    <span
      class="d-flex align-items-center"
    >
      -
    </span>
    <button
      class="btn btn-outline-dark btn-sm w-100 ms-1"
      data-testid="shortcut-today-to-1720000034990"
      type="button"
    >
      Today
    </button>
  </div>
</div>
`;
